kpi-tracker/src/main.rs

```diff
@@ -0,0 +1,277 @@
+        BlockItemSummaryDetails::AccountCreation, CredentialType,
+    },
+    v2::{self, Client, Endpoint},
+};
+use futures::{self, StreamExt, TryStreamExt};
+
+#[derive(Debug, Parser)]
+struct Args {
+    /// The node used for querying
+    #[arg(
+        long = "node",
+        help = "The endpoint is expected to point to a concordium node grpc v2 API.",
+        default_value = "http://localhost:20001"
+    )]
+    node: Endpoint,
+    /// How many blocks to process.
+    // Only here for testing purposes...
+    #[arg(long = "num-blocks", default_value_t = 10000)]
+    num_blocks: u64,
+}
+
+/// Holds information about accounts created on chain.
+#[derive(Debug)]
+struct AccountDetails {
+    /// Whether an account was created as an initial account or not.
+    is_initial: bool,
+    /// Link to the block in which the account was created.
+    block_hash: BlockHash,
+}
+
+/// Information about individual blocks. Useful for linking entities to a block and it's
+/// corresponding attributes.
+#[derive(Debug, Clone, Copy)]
+struct BlockDetails {
+    /// Finalization time of the block. Used to show how metrics evolve over time by linking entities, such as accounts and transactions, to
+    /// the block in which they are created.
+    block_time: DateTime<Utc>,
+    /// Height of block from genesis. Used to restart the process of collecting metrics from the
+    /// latest block recorded.
+    height: AbsoluteBlockHeight,
+}
+
+type AccountsTable = HashMap<AccountAddress, AccountDetails>;
+type BlocksTable = HashMap<BlockHash, BlockDetails>;
+
+/// This is intended as a in-memory DB, which follows the same schema as the final DB will follow.
+struct DB {
+    /// Table containing all blocks queried from node.
+    blocks: BlocksTable,
+    /// Table containing all accounts created on chain, along with accounts present in genesis
+    /// block
+    accounts: AccountsTable,
+}
+
+/// Queries node for account info for the `account` given at the block represented by the
+/// `block_hash`
+fn account_details(
+    block_hash: BlockHash,
+    account_creation_details: Option<AccountCreationDetails>,
+) -> AccountDetails {
+    let is_initial = account_creation_details.map_or(false, |act| match act.credential_type {
+        CredentialType::Initial { .. } => true,
+        CredentialType::Normal { .. } => false,
+    });
+
+    AccountDetails {
+        is_initial,
+        block_hash,
+    }
+}
+
+/// Returns accounts on chain at the give `block_hash`
+async fn accounts_in_block(
+    node: &mut Client,
+    block_hash: BlockHash,
+) -> anyhow::Result<BTreeMap<AccountAddress, AccountDetails>> {
+    let accounts = node
+        .get_account_list(block_hash)
+        .await
+        .context(format!("Could not get accounts for block: {}", block_hash))?
+        .response;
+
+    let accounts_details_map = accounts
+        .map_ok(|account| {
+            let details = account_details(block_hash, None);
```

Comment:
Isn't this equivalent to saying `let details = AccountDetails{is_initial: false, block_hash}`?

Which is not exactly correct.

Reply:
Yes it is, and not it's not correct in a general sense. It was intended for genesis accounts only, but isn't really apparent in any way. I'll restructure so it becomes clear.